common/src/api/internal/nexus.rs

```diff
@@ -29,40 +29,65 @@ pub struct DiskRuntimeState {
     pub time_updated: DateTime<Utc>,
 }
 
-/// Runtime state of the Instance, including the actual running state and minimal
-/// metadata
-///
-/// This state is owned by the sled agent running that Instance.
+/// The "static" properties of an instance: information about the instance that
+/// doesn't change while the instance is running.
 #[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
-pub struct InstanceRuntimeState {
-    /// runtime state of the Instance
-    pub run_state: InstanceState,
-    /// which sled is running this Instance
-    pub sled_id: Uuid,
-    /// which propolis-server is running this Instance
-    pub propolis_id: Uuid,
-    /// the target propolis-server during a migration of this Instance
-    pub dst_propolis_id: Option<Uuid>,
-    /// address of propolis-server running this Instance
-    pub propolis_addr: Option<SocketAddr>,
-    /// migration id (if one in process)
-    pub migration_id: Option<Uuid>,
-    /// The generation number for the Propolis and sled identifiers for this
-    /// instance.
-    pub propolis_gen: Generation,
-    /// number of CPUs allocated for this Instance
+pub struct InstanceProperties {
     pub ncpus: InstanceCpuCount,
-    /// memory allocated for this Instance
     pub memory: ByteCount,
-    /// RFC1035-compliant hostname for the Instance.
+    /// RFC1035-compliant hostname for the instance.
     // TODO-cleanup different type?
     pub hostname: String,
-    /// generation number for this state
+}
+
+/// The dynamic runtime properties of an instance: its current VMM ID (if any),
+/// migration information (if any), and the instance state to report if there is
+/// no active VMM.
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+pub struct InstanceRuntimeState {
+    /// The state of the instance if it has no active VMM.
+    pub fallback_state: InstanceState,
```

Comment:
We talked about this some offline (and I discussed it some in another comment). I think the ideal semantics are

- when there's an VMM, sled agent owns the instance's logical state and puts it in the active VMM's record
- when there's no VMM, Nexus owns the instance's logical state and puts it in the instance record

Right now this breaks down a bit because sled agent updates the fallback state when an active Propolis is retired. We could change that a bit--if Nexus sees that the active Propolis has gone away, then it can insert the appropriate state into the instance record before writing anything back to CRDB.

I'm going to think about this one a bit more/sleep on it and then decide how to move forward.

Reply:
Fixed in d2a063668: the `fallback_state` field is now named `nexus_state`, and Nexus is the only thing that modifies it (enforced by removing the corresponding field from the structs Nexus and sled agent send each other). The instance state machine should now reduce to the following:

- Creating: this instance record is still being set up
- Stopped: this instance has no active VMM and so isn't running anywhere
- Failed: similar to Stopped except that Nexus thinks the instance had a Problem
- Running: the instance has an active VMM; go look at it for its state
- Destroyed: the instance is gone (it should also have a non-NULL `time_deleted`)

Instead of getting the Running -> Stopped transition from sled agent, Nexus recomputes the effective state when applying sled agent state updates based on whether the updated runtime states have a `propolis_id` or not. (This is the computation sled agent was doing; it's just happening up in Nexus now.)

This was a surprisingly small change (35 added, 68 removed, and a lot of that was just renaming), which makes me more confident it's correct. I also audited the code for places where we assign the Nexus state and verified that only the states I listed above are actually used.

> Does it make sense to have this be a more-constrained set of states than InstanceState?

Yes, we should enforce this whole thing by constraining the values this column can have. I'd like to do that in a separate PR because I expect it to be more tedious (we'll need to figure out the schema update and then create/plumb in all the new Rust types).

Reply:
This is great, thank you for taking the time to do this pass. I like this new version a lot!